lib/quicrq.c

```diff
@@ -1828,22 +1913,41 @@ int quicrq_callback(picoquic_cnx_t* cnx,
         case picoquic_callback_stream_data:
         case picoquic_callback_stream_fin:
             /* Data arrival on stream #x, maybe with fin mark */
-            if (stream_ctx == NULL) {
-                /* Retrieve, or create and initialize stream context */
+            if ((stream_id & 2) == 0 && stream_ctx == NULL) {
+                /* Retrieve, or create and initialize stream context for control channel */
                 stream_ctx = quicrq_find_or_create_stream(stream_id, cnx_ctx, 1);
-            }
+                if (stream_ctx == NULL) {
+                    /* Internal error */
+                    (void)picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return(-1);
+                }
 
-            if (stream_ctx == NULL) {
-                /* Internal error */
-                (void)picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
-                return(-1);
-            }
-            else {
-                ret = quicrq_receive_stream_data(stream_ctx, bytes, length, (fin_or_event == picoquic_callback_stream_fin));
+                /* we consider unidir streams as warp/rush streams. may be we need something more explicit */
+                ret = quicrq_receive_stream_data(stream_ctx, bytes, length,(fin_or_event == picoquic_callback_stream_fin));
+
+            } else {
+                /* uni-directional stream are used for warp/rush flows */
+                if (cnx_ctx->control_stream == NULL) {
+                    /* Internal error, control stream should have been set up */
+                    (void) picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return (-1);
+                }
+
+                uni_stream_ctx = quicrq_find_or_create_uni_stream(stream_id, cnx_ctx, 1);
+                if (uni_stream_ctx == NULL) {
+                    /* Internal error */
+                    (void) picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return (-1);
+                }
+
+                /* we consider unidirectional streams as warp/rush streams. maybe we need something more explicit */
+                ret = quicrq_receive_warp_or_rush_stream_data(uni_stream_ctx, bytes, length,
+                                                              (fin_or_event == picoquic_callback_stream_fin));
             }
+
             break;
         case picoquic_callback_prepare_to_send:
-            /* Active sending API */
+            /* Active sending API, called after wake up and picoquic conditions (congestion/rate control are ready to be sent */
             if (stream_ctx == NULL) {
                 /* This should never happen */
                 picoquic_log_app_message(cnx, "QUICRQ callback returns %d, event %d", ret, fin_or_event);
```

Comment:
TODO: here too you have to plug in the uni_stream specific function for sending data. Check that there is still something to send in the GOP, etc.